### ISSUE TITLE
RootState typing

### ISSUE THREAD

**Issue by hachibeeDI** [jonasbuntinx/react-boilerplate#2]
Whether we are using combine/reducer reducer, RootState/AppState should be generated automatically.
I mean, declaring `export interface State {` in each feature as Store typing may could be maintain cost which may introduce human error.

TypeScript supports kinda use case via `infer`.
I think we can do something like below:

```typescript
// src/store.ts
const rootReducer = combineReducers({
  app: reducer,
  router: routerReducer,
  login: loginReducer,
  registration: registrationReducer
});

export const store = createStore(rootReducer);
```

```typescript
// src/redux/types.ts
import {store} from '~/src/store';

export type RootState = ReturnType<typeof store.getState>;
```

```typescript
// src/feature/something/index.tsx

export connect((state: RootState): FeatureState => ({ // ...
```

I guess we can't go straight with my idea though, defining RootState will be useful if we gonna migrate to flatten reducer. 🤔

**Comment by jonasbuntinx** [comment]
Yeah, I deliberately avoided creating a `RootState` so that developers have to thinks about what state is being used in the `connect` function. 

Without a `RootState` type you are forced to pattern match the state slices you want to use. In most cases this will only be the app state and their own feature state.

```
// src/feature/something/index.tsx

export connect((state: { app: AppState; feature: FeatureState }): FeatureState => ({ // ...
``` 

But maybe this is overkill. 🤔

**Comment by hachibeeDI** [comment]
Yeah, consider what is going to picked is good design manner. 👌 
I think defining `FeatureState` is the progress of consideration. `state: Root/AppState` is a fact how application being so we can keep it as it is.

And if I understand correctly, we need RootState after migrate to reduceReducer because all of reducer signatured as `(state: RootState, action: Action) => RootState`?

**Comment by jonasbuntinx** [comment]
Yes, if we use reduceReducer we do need a RootState.

But currently I stepped away from using reduceReducer as it might become too difficult to migrate to.
Also reduceReducer allows features to change global state directly. After consideration I believe that should not be allowed.

I'm going to look more into typeless.js. If this comes close to my version of handling redux, then maybe it is beneficial to use it instead of rolling out our own solution.

**Comment by hachibeeDI** [comment]
That's make sense & Sounds good!

For me typeless is more looks like Elm Architecture than Redux. And I'm sure it will introduce typefull developer experience without bunch of weird plugins or knee deep TypeScript knowledge. :)
